File: 1046. Last Stone Weight/1046. Last Stone Weight.py
import heapq
from typing import List


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        for i in range(len(stones)):
            stones[i] *= -1

        heapq.heapify(stones)

        while len(stones) > 1:
            largest = heapq.heappop(stones)
            second_largest = heapq.heappop(stones)

            if largest != second_largest:
                heapq.heappush(stones, largest - second_largest)

        return -heapq.heappop(stones) if stones else 0

File: 1046. Last Stone Weight/1046. Last Stone Weight.js
/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function (stones) {
  const maxHeap = new MaxPriorityQueue();

  for (stone of stones) {
    maxHeap.enqueue(stone);
  }

  while (maxHeap.size() > 1) {
    let largest = maxHeap.dequeue();
    const secondLargest = maxHeap.dequeue();

    if (largest !== secondLargest) {
      largest -= secondLargest;
      maxHeap.enqueue(largest);
    }
  }

  return maxHeap.size() === 0 ? 0 : maxHeap.front();
};
